### ISSUE TITLE
Question on Mendler-style recursion

### ISSUE THREAD

**Issue by atennapel** [cryptiumlabs/juvix#301]
In the language reference in 8.3 it notes that a function `out : FixM f -> f (FixM f)` is required to write `pred`. Some questions:

- How would one write `pred` if `out` was available? In the algebra the predecessor of `S` is of the abstract type `r`, so how would `out` help at all? Don't we need a function from `r` to `FixM Nat`?
- Could `out` be a constant time primitive?
- Could a primitive `Fix` type solve the issue?

**Comment by mariari** [comment]
Hello atennapel, thank you for asking. 

It's been a while since I wrote that section, so if there are some minor inaccuracies in my response, that is why. 

> How would one write pred if out was available?

[I have some LISP code, that shows how out works](https://github.com/cryptiumlabs/juvix/blob/develop/experimental/playground/mendler.lisp)

```lisp
(defun out (d)
  (funcall d
           (lambda (rec fr) (fmap (lambda (r) (fn:curry in (funcall rec r))) fr))))

;; first attempt O(1)
(defun pred-alg% (n)
  (let ((var (out n)))
    (print "time")
    (if (equalp +Z+ var)
        #'zero%
        (s-param var))))
```

The O(n) is quite noticeable when we define `hundred%` as follows
```lisp
(let ((reduced
        (reduce (lambda (x acc)
                  (declare (ignore x))
                  (fn:curry succ% acc))
                (list:range 0 12000)
                :initial-value #'zero%
                :from-end t)))
  (defun hundred% (x)
    (succ% reduced x)))

(time (pred-alg% #'hundred%))
(time (hundred% #'pred-alg))
```

There is also [A haskell version here](https://github.com/cryptiumlabs/juvix/blob/develop/experimental/playground/Mendler.hs)

```haskell
out'' ∷ ∀ f. Functor f ⇒ (FixM f) → f (FixM f)
out'' = undefined
```
From this type we can See that we go from FixM Nat → Nat (FixM Nat)

> In the algebra the predecessor of S is of the abstract type r, so how would out help at all? Don't we need a function from r to FixM Nat?

I think an expansion would give some intuition on how out gives back the correct type

```lisp
out = λd. d (λr f. fmap (λa. in (r a)) f)

in Z = λf. f (λd. d f) Z

in (S (in z)) = λ f. f (λ d. d f) (S (in Z))

;; this is short hand!
#1 = (λr f. fmap (λa. in (r a)) f)

;; Note that we will later change the (in Z) to (S (in z)) near
;; the end so we can show both cases!
out (in Z)
= out (λf. f (λd. d f) Z)
= (λr f. fmap (λ a. in (r a)) f) (λ d. d #1) Z
= fmap (λa. in ((λ d. d #1) a)) Z
= fmap (λa. in (a #1)) Z
= Z ; Definition of fmap f Z = Z

;; Now let's suppose we had (S (in Z)) instead of Z now!
= fmap (λ a. in (a #1)) (S (in Z))
= S ((λa. in (a #1)) (in Z)) ; definition of fmap S
= S (in ((in Z) #1))
= S (in Z)                   ; we know this by the end of the Z computation!
```

As for the other two questions, I will let my colleagues answer those.

**Comment by atennapel** [comment]
Hi, thanks for your answer.

Let me rephrase the question:
```
-- defining the pred mendler algebra
NatF r = ZF | SF r
Nat = FixM NatF

pred : AlgM NatF Nat (= forall r. (r -> Nat) -> NatF r -> Nat)
pred rec ZF = Z
pred rec (SF x) = x
-- semantically x is actually a Nat, but we only know that x is of type r,
-- so we cannot actually return it here.
-- having an out function does not help here!
```

**Comment by cwgoes** [comment]
> So the constant time `out` can be defined as `out = \f. f \x y. y`, I think the "Efficient Mendler-Style Lambda-Encodings in Cedille" just manages to actually make this typecheck consistently.

Yes, exactly.

> In my system I don't care much about consistency (I already have `Type : Type`) so I could define a primitive `out` as `\f. f \x y. y` with the type `{f : Type -> Type} -> Fix f -> f (Fix f)` and get constant-time deconstructors.

I don't know your system in detail but if you aren't worried about consistency that seems OK; in the untyped version `pred` can be written in the same form as `is-even`, just ignoring the `rec` argument (see [Lisp example](https://github.com/cryptiumlabs/juvix/blob/develop/experimental/playground/mendler.lisp#L29)).

**Comment by atennapel** [comment]
Thanks!

I have implemented it and it works: https://github.com/atennapel/coc-os/blob/master/lib/mendler.p
